### PR TITLE
HypreParMatrix Frobenius Norm

### PR DESCRIPTION
**Pull request by thartland**
Adding the wrapper HypreParMatrix::FNorm of [hypre_PARCSRMatrixNormFro](https://github.com/hypre-space/hypre/blob/master/src/parcsr_ls/par_ilu.c#L3921) to compute the Frobenius norm of a HypreParMatrix.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4548](https://github.com/mfem/mfem/pull/4548) | @thartland | @pazner | @dylan-copeland + @pazner | 10/30/24 | 11/1/24 | 11/12/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by pazner** [comment]
Notation should probably be consistent with `DenseMatrix::FNorm()` (or vice versa).

**Comment by thartland** [comment]
Thanks @pazner for catching that.

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by pazner — APPROVED** [review]
Thanks @thartland

**Comment by pazner** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
We get a failure with hypre v2.10 (which we use in some of the tests):
```
linalg/hypre.cpp(2801): error: identifier "hypre_ParCSRMatrixNormFro" is undefined
        const int ierr = hypre_ParCSRMatrixNormFro(A, &norm_fro);
                         ^

1 error detected in the compilation of "linalg/hypre.cpp".
```
This method was added in hypre v2.19.0, so we should add an alternative implementation for older versions, e.g. based on `Vector::Norml2` for the diagonal and off-diagonal blocks wrapped as `Vector`s.

File: linalg/hypre.cpp [hunk 1]
```diff
@@ -2780,11 +2780,21 @@ void HypreParMatrix::PrintHash(std::ostream &os) const
...
+if (A != NULL)
+#if MFEM_HYPRE_VERSION >= 21900
    {
       const int ierr = hypre_ParCSRMatrixNormFro(A, &norm_fro);
       MFEM_VERIFY(ierr == 0, "");
    }
+#else
+   {   
+      MPI_Comm comm = hypre_ParCSRMatrixComm(A);
+      hypre_CSRMatrix *hypre_merged = hypre_MergeDiagAndOffd(A);
```
Comment (v-dobrev) [inline linalg/hypre.cpp:2792]: Merging the diagonal and off-diagonal blocks is not a cheap operation. It is better to compute the norms of the blocks separately and then combine them.

File: linalg/hypre.cpp [hunk 2]
```diff
@@ -2777,6 +2777,34 @@ void HypreParMatrix::PrintHash(std::ostream &os) const
...
+      normsqr_fro_loc += InnerProduct(Avec_diag, Avec_diag);
+      csr = A->offd;
+      Vector Avec_offd(csr->data, csr->num_nonzeros);
+      normsqr_fro_loc += InnerProduct(Avec_offd, Avec_offd);
+      MPI_Allreduce(&normsqr_fro_loc, &norm_fro, 1, MPITypeMap<real_t>::mpi_type, MPI_SUM, comm);
+      norm_fro = sqrt(norm_fro);
+   }
+#endif
+   return norm_fro;
+}
```
Comment (pazner) [inline linalg/hypre.cpp:2805]: Minor style/formatting (the style CI check is failing)

```suggestion
real_t HypreParMatrix::FNorm() const
{
   real_t norm_fro = 0.0;
   if (A != NULL)
#if 0
   {
      const int ierr = hypre_ParCSRMatrixNormFro(A, &norm_fro);
      MFEM_VERIFY(ierr == 0, "");
   }
#else
   {
      Vector Avec_diag(A->diag->data, A->diag->num_nonzeros);
      real_t normsqr_fro = InnerProduct(Avec_diag, Avec_diag);
      Vector Avec_offd(A->offd->data, A->offd->num_nonzeros);
      normsqr_fro += InnerProduct(Avec_offd, Avec_offd);
      MPI_Allreduce(MPI_IN_PLACE, &normsqr_fro, 1, MPITypeMap<real_t>::mpi_type,
                    MPI_SUM, hypre_ParCSRMatrixComm(A));
      norm_fro = sqrt(normsqr_fro);
   }
#endif
   return norm_fro;
}
```

**Comment by pazner** [comment]
@v-dobrev, would this be the right way of handling the case where the hypre matrix is stored on device?

```c++
Vector Avec;
Avec.NewMemoryAndSize(mem_diag.data, A->diag->num_nonzeros, false);
real_t normsqr_fro = InnerProduct(Avec, Avec);
Avec.NewMemoryAndSize(mem_offd.data, A->offd->num_nonzeros, false);
normsqr_fro += InnerProduct(Avec, Avec);
```

**Comment by v-dobrev** [comment]
@pazner, for hypre built with GPU support we require versions newer than 2.19.0 (see the `INSTALL` file), so the code for hypre < 2.19.0 can just assume hypre is built for CPU only.

**Comment by pazner** [comment]
OK, sounds good. @thartland, could you add a comment to the code mentioning that in that case the matrix is guaranteed to be in host memory?

**Comment by thartland** [comment]
Thanks for asking about this @pazner! The requested comment has been added.

**Review by pazner — APPROVED** [review]

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing...

**Pull request merged**